segmentstore/server/src/main/java/io/pravega/segmentstore/server/containers/StorageEventProcessor.java

```diff
@@ -0,0 +1,141 @@
+import io.pravega.segmentstore.storage.chunklayer.AbstractTaskQueueManager;
+import io.pravega.segmentstore.storage.chunklayer.GarbageCollector;
+import lombok.Getter;
+import lombok.extern.slf4j.Slf4j;
+import lombok.val;
+
+import java.io.IOException;
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * Implementation of {@link AbstractTaskQueueManager} that uses {@link io.pravega.segmentstore.server.ContainerEventProcessor.EventProcessor}
+ * as underlying implementation.
+ * This class acts as adaptor that converts calls on {@link io.pravega.segmentstore.server.ContainerEventProcessor.EventProcessor}
+ * into appropriate calls on {@link AbstractTaskQueueManager} and vice versa.
+ */
+@Slf4j
+public class StorageEventProcessor implements AbstractTaskQueueManager<GarbageCollector.TaskInfo> {
+
+    private final static GarbageCollector.TaskInfo.Serializer SERIALIZER = new GarbageCollector.TaskInfo.Serializer();
+
+    private final int containerID;
+    private final ContainerEventProcessor eventProcessor;
+    private final Function<List<GarbageCollector.TaskInfo>, CompletableFuture<Void>> callBack;
+    private final int maxItemsAtOnce;
+    private final String traceObjectId;
+    @Getter
+    private final ConcurrentHashMap<String, ContainerEventProcessor.EventProcessor> eventProcessorMap = new ConcurrentHashMap<>();
+
+    /**
+     * Constructor.
+     * @param containerID Container id.
+     * @param eventProcessor Instance of {@link ContainerEventProcessor} to use.
+     * @param callBack Function which is called to process batch of events.
+     * @param maxItemsAtOnce Maximum nuber of
+     */
+    public StorageEventProcessor(int containerID,
+                                 ContainerEventProcessor eventProcessor,
+                                 Function<List<GarbageCollector.TaskInfo>,
+                                 CompletableFuture<Void>> callBack,
+                                 int maxItemsAtOnce) {
+        this.containerID = containerID;
+        this.eventProcessor = Preconditions.checkNotNull(eventProcessor, "eventProcessor");
+        this.callBack = Preconditions.checkNotNull(callBack, "callBack");
+        this.maxItemsAtOnce = maxItemsAtOnce;
+        this.traceObjectId = String.format("StorageEventProcessor[%d]", containerID);
+    }
+
+    /**
+     * Adds a queue by the given name.
+     *
+     * @param queueName Name of the queue.
+     * @param ignoreProcessing Whether the processing should be ignored.
+     */
+    @Override
+    public CompletableFuture<Void> addQueue(String queueName, Boolean ignoreProcessing) {
+        Preconditions.checkNotNull(queueName, "queueName");
+        val config = new ContainerEventProcessor.EventProcessorConfig(maxItemsAtOnce, Long.MAX_VALUE);
+        val f = ignoreProcessing ?
+                eventProcessor.forDurableQueue(queueName) :
```

Comment:
These calls can fail, for instance if the internal segment is not yet ready or any other temporary issue. What is the expected behavior if it does fail? Is this going to retry, it will shut down the container or it will just carry on? Because in the latter case, we may face a situation in which the Segment Store works but the `EventProcessor`s are not initialized, so garbage would not be collected, right? I say that not as a theoretical scenario, but I remember seeing this problem in one of the initial longevity runs in which I induced restarts in purpose (the code may have changed since then, but we need tests to validate this case).

Reply:
This method is called during container bootstrap/startup.
As of today we do not retry individual steps in container startup.
In case of failure the container will fail to start and go through re-initialization (similar to when other components fail during startup).